### PR TITLE
Collapse Activities in navbar based on avail space (& preference)

### PR DESCRIPTION
**Pull request by cpeel**
Introduce the idea of drop-down menu options to group like things together and make better use of the top header space. Much to my chagrin we use the new menu based on a user preference; if the preference is set to "auto" (the default) dynamically use the menu based on the width of the window.

This is available for testing in https://www.pgdp.org/~cpeel/c.branch/refine-navbar/

See also the discussion in the [Proposed changes to the navigation bar - 2024 edition](https://www.pgdp.net/phpBB3/viewtopic.php?f=4&t=81413) forum.

Testing notes: In addition to confirming this works in all our minimum browsers, we should make sure this looks good in all themes -- including having an unread message in the forums which will add an "unread messages" number to the new Inbox icon -- and on various mobile devices.

### CODE REVIEW

File: pinc/theme.inc [hunk 1]
```diff
@@ -119,30 +119,96 @@ function html_logobar()
...
+    global $blog_url, $wiki_url;
 
     echo "\n<div id='header'>\n";
-    echo "<a href='$code_url'>\n";
+
     echo "<div id='logo'>\n";
-    echo "<img src='$code_url/graphics/dp-logo.svg' alt='$site_name'>\n";
-    echo "</div>\n";
+    echo "<a href='$code_url'>\n";
+    echo "<img id='logo-img' src='$code_url/graphics/dp-logo.svg' alt='$site_name'>\n";
```
Comment (bpfoley) [inline pinc/theme.inc:128]: I wonder do we want to escape $site_name? It could contain single quotes, e.g. "Brian's Proofing Emporium".

File: pinc/theme.inc [hunk 2]
```diff
@@ -119,30 +119,96 @@ function html_logobar()
...
+    echo "</div>\n"; // logo
 
-    echo "<div id='titles-preserved'><p>\n";
-
-    // show number of books posted to PG
-    // (takes into account that some books are processed as several projects)
-    echo "<span id='x-preserved'>";
-    echo sprintf(
-        _('%s titles preserved for the world!'),
-        number_format(total_completed_projects())
```
Comment (bpfoley) [inline pinc/theme.inc:137]: Saves an extra trip to the DB. Woo.

File: pinc/theme.inc [hunk 3]
```diff
@@ -119,30 +119,96 @@ function html_logobar()
...
+        echo "<span id='month-preserved-summary'>";
+        show_completed_projects(true);
+        echo "</span>";
+        echo "</div>\n";
+    } else {
+        $icon_links = [];
+        if (get_url_for_forum()) {
+            $icon_links[] = [
+                "name" => _("Inbox"),
+                "class" => "fas fa-envelope fa-2x",
```
Comment (bpfoley) [inline pinc/theme.inc:153]: Perhaps pull "fa-2x" out as something that's common to all the icons?

File: pinc/theme.inc [hunk 4]
```diff
@@ -230,29 +299,59 @@ function html_navbar()
...
 
         $links = [];
         $links[] = ['text' => _("Activity Hub"), 'url' => "activity_hub.php"];
 
-        // use white-space: nowrap; to keep the rounds within []s together
-        $activities_string = "<span style='white-space: nowrap;'>[ ";
-        if (get_pages_proofed_maybe_simulated() < 100) {
-            $activities_string .= "<span class='text'>" . _("Activities") . ": </span>";
+        // We create 2 different ways of showing the activities, one with them
+        // all out in one string and one with with them collapsed behind a menu.
```
Comment (bpfoley) [inline pinc/theme.inc:314]: s/all out/all visible/?

File: pinc/theme.inc [hunk 5]
```diff
@@ -230,29 +299,59 @@ function html_navbar()
...
+        // Furthermore, never collapse the menu for new users or users with
+        // fewer than 4 activities.
+        $activity_links = get_activity_links();
+
+        $is_new_user = get_pages_proofed_maybe_simulated() < 100;
+
+        // Calculate style-level CSS that will override the classes
+        // 2 - never collapse
+        if ($user->navbar_activity_menu == 2 || $is_new_user || count($activity_links) <= 3) {
+            $string_style = "style='display: inline;'";
```
Comment (bpfoley) [inline pinc/theme.inc:328]: You can drop all the trailing semicolons in your HTML `style='foo: bar;'` attributes.

File: pinc/theme.inc [hunk 6]
```diff
@@ -270,43 +369,52 @@ function html_navbar()
...
-        if (get_url_for_forum()) {
-            $links[] = ['text' => _("Forums"), 'url' => get_url_for_forum()];
-        }
+        echo headerbar_text_array($links);
+        echo "  &#183; ";
 
-        if (!empty($blog_url)) {
-            $links[] = ['text' => _("Blog"), 'url' => $blog_url];
-        }
+        $account_links = [];
```
Comment (bpfoley) [inline pinc/theme.inc:377]: Maybe just make this a single assign rather than appending? 
```
$account_links = [ ..., ... , ... ];
```

File: pinc/theme.inc [hunk 7]
```diff
@@ -270,43 +369,52 @@ function html_navbar()
...
+{
+    $menu_style = "";
+    $menu_contents_style = "";
+    if ($align_right) {
+        $menu_style = "style='padding-right: 0.5em;'";
+        $menu_contents_style = "style='right: 0;'";
+    }
+
+    $menu_id = 'menu' . md5($menu_title);
+    $output = [];
```
Comment (bpfoley) [inline pinc/theme.inc:401]: Again, perhaps single assign here rather than repeated append. Makes it a little less noisy.

File: scripts/navbar_menu.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+document.addEventListener("click", function (event) {
+    let inMenu = false;
+    document.querySelectorAll(".menu").forEach(element => {
+        if (element.contains(event.target) || element == event.target) {
+            inMenu = true;
+        }
+    });
+
+    if (!inMenu) {
+        document.querySelectorAll(".menu-contents").forEach(element => {
```
Comment (bpfoley) [inline scripts/navbar_menu.js:25]: This duplicates the actions of `toggleMenu`. Can we reword this as something like
```
document.querySelector(".menu").forEach(element => toggleMenu(element.id));
```

File: styles/layout.less [hunk 9]
```diff
@@ -219,13 +220,43 @@ header, #header {
...
+        .center-align;
+        text-decoration: none;
+        .sans-serif;
+        font-size: 0.7em;
+        color: @heading-color;
+        .unicolor-link(@heading-color);
+        .icon-menu-item-name {
+            font-size: 0.9em;
+        }
+        .inbox-badge {
```
Comment (bpfoley) [inline styles/layout.less:247]: I feel that the badge is slightly lower than looks good. Any chance we can push it up a little bit?

File: styles/layout.less [hunk 10]
```diff
@@ -219,13 +220,43 @@ header, #header {
...
+          background-color: @text-error;
+          border-radius: 2px;
+          color: @page-background;
+
+          padding: 1px 3px;
+          font-size: 10px;
```
Comment (bpfoley) [inline styles/layout.less:253]: Should `font-size` be expressed in `em`s rather than as an absolute `px` size?

File: pinc/theme.inc [hunk 11]
```diff
@@ -230,29 +299,59 @@ function html_navbar()
...
         // left side
         echo "<span id='navbar-left'>";
-        echo headerbar_text_array([['text' => $site_abbreviation, 'url' => "default.php"]]);
-
-        echo " | ";
 
         $links = [];
         $links[] = ['text' => _("Activity Hub"), 'url' => "activity_hub.php"];
 
-        // use white-space: nowrap; to keep the rounds within []s together
```
Comment (bpfoley) [inline pinc/theme.inc:248]: Note to self: `white-space: nowrap` has moved into the CSS files.

**Review by bpfoley — APPROVED** [review]

File: scripts/navbar_menu.js (hunk 8 above)
Comment (cpeel) [reply scripts/navbar_menu.js:25]: Ah, but it doesn't actually. `toggleMenu()` just toggles. If we called `toggleMenu()` it would close any open menus and open any closed menus, it wouldn't close everything. If there were 1 and only 1 menu they would be the same, but with more than 1 we can't use it.

File: pinc/theme.inc (hunk 3 above)
Comment (cpeel) [reply pinc/theme.inc:153]: Fair question, but to me the whole class attribute feels like a unit here and I'm wary of breaking that into two pieces, one of which is stashed out in a stylesheet.

File: pinc/theme.inc (hunk 5 above)
Comment (cpeel) [reply pinc/theme.inc:328]: Noted, but I think I'll leave them (makes it easier to add new ones if necessary, move it to stylesheet if necessary, and I'm lazy 😁 ).

**Review by cpeel — COMMENTED** [review]
Will address (eg make code changes) on your other items later today. Thanks for the review!

File: scripts/navbar_menu.js (hunk 8 above)
Comment (bpfoley) [reply scripts/navbar_menu.js:25]: Ah. Yeah, I didn't read closely enough. Good point!

File: scripts/navbar_menu.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+            element.classList.add('invisible');
+            element.classList.add('transparent');
```
Comment (chrismiceli) [inline scripts/navbar_menu.js:27]: does this work for things like preventing tabbing to these elements?

**Comment by windymilla** [comment]
Tested in Chrome Version 123.0.6312.105 (Official Build) (64-bit) on Windows:
All behaved as expected, including with extremely small/large font sizes, and with all 3 prefs settings.

Love the horizontal drop-down, and the little arrow works for both opening & closing the dropdown.

**Review by windymilla — APPROVED** [review]
It seems odd that there is now a pref that will have no effect for a new user. If a new user is exploring and changes that pref to Always or Never, perhaps that should be respected. I think it should only have the "new user" override when it is set to auto. For a new user "Auto" means "Always". Just a small thing

Congrats on tracking through all the places that needed edits to implement this!

File: pinc/theme.inc (hunk 1 above)
Comment (cpeel) [reply pinc/theme.inc]: Good catch - fixed this and the other instance in this file.

File: pinc/theme.inc (hunk 6 above)
Comment (cpeel) [reply pinc/theme.inc]: Oh yeah, that's much better. And obvious in retrospect.

File: pinc/theme.inc (hunk 7 above)
Comment (cpeel) [reply pinc/theme.inc]: Yup, done.

File: scripts/navbar_menu.js [hunk 13]
```diff
@@ -0,0 +1,33 @@
+/* exported toggleMenu */
+
+// Toggle a menu contents to visible and rotate the menu icon
+function toggleMenu(menuId) {
+    let menu = document.getElementById(menuId);
+    [menu.getElementsByClassName("menu-contents")].forEach(element => {
```
[the remark this answers is not in the document]
Comment (cpeel) [reply scripts/navbar_menu.js:6]: I admit I don't know the difference. What I have works but I was literally throwing shit at the wall. Would you mind showing me what you think this should look like? I'm not familiar with the `[...` syntax and don't know what it's called to look it up.

File: scripts/navbar_menu.js (hunk 12 above)
Comment (cpeel) [reply scripts/navbar_menu.js:27]: I don't know. In the browser how do you access something via tab that isn't an input? I'm using Firefox and clicking into the window and hitting tab isn't letting me navigate around.

**Review by cpeel — COMMENTED** [review]
> It seems odd that there is now a pref that will have no effect for a new user. If a new user is exploring and changes that pref to Always or Never, perhaps that should be respected. I think it should only have the "new user" override when it is set to auto. For a new user "Auto" means "Always".

That's a really great point, thanks for bringing it up. I've updated so that we only override the collapsing in "auto" mode.

---

The last commit doesn't resolve all the comments, more to come.

File: scripts/navbar_menu.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (chrismiceli) [reply scripts/navbar_menu.js:6]: I was actually slightly wrong, you don't even need the `...` in this case.

The return type from getElementsByClassName does not have a `forEach` method as it is not an Array type for some strange reason. That means if you want to go through each element returned, you traditionally have to do a silly old for loop:
```
const menuContents = menu.getElementsByClassName('menu-contents');
for (let i = 0; i < menuContents.length; i++)
{
  const myElement = menuContents.item(i);
  // ...
}
```

Obviously this is kind of verbose. You can however use its [iterator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/iterator) properties to loop through
```
const menuContentsIterator = menu.getElementsByClassName('menu-contents')[Symbol.iterator]().next();
while(menuContentsIterator.done === false)
{
  const myElement = menuContentsIterator.value;
  // ...
  menuContentsIterator = menuContentsIterator.next();
}
```
Now that is not much better, so there is some syntatic sugar to use the iterator with a for of loop:
```
for (const myElement of menu.getElementsByClassName('menu-contents'))
{
  // ...
}
```

File: scripts/navbar_menu.js (hunk 12 above)
Comment (chrismiceli) [reply scripts/navbar_menu.js:27]: I was referring to child elements (anchors), but it seems you set them to `visibility: hidden`, so that should make them not tabbable.

File: scripts/navbar_menu.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (cpeel) [reply scripts/navbar_menu.js:6]: That's perfect, Chris - thank you!

**Comment by cpeel** [comment]
Latest commits do the following:
* Updates how we're iterating over the elements in the JS (thanks Chris!)
* Changes the icon colors back to the navbar background (except for Charcoal) -- the text remains the header color which is every so slightly darker (in most cases) but isn't black.
* Addresses other PR requests, including moving the unread messages slightly up and to the right from where it was.

Finally, it includes a separate commit that will allow someone to use the `?numofpages=` parameter on TEST to set it to any value, not just lower like it does on PROD.

I'll rebase the fixup commits after we've finished iterating over the changes in the PR & testing.

**Comment by windymilla** [comment]
> Finally, it includes a separate commit that will allow someone to use the `?numofpages=` parameter on TEST to set it to any value, not just lower like it does on PROD.

Thanks - just returned a page to the rounds so I'm back down to 99 P1 pages on TEST, and confirmed `numofpages` param worked correctly for me.

**Review by 70ray — APPROVED** [review]

**Review by chrismiceli — APPROVED** [review]

**Comment by cpeel** [comment]
Some browser testing results, all good.

# Our minimum supported versions
Chrome 57 on Windows 10
Firefox 52 on Windows 10
Opera 44 on Windows 10
Safari 11 on High Sierra

# Latest versions
Chrome: 123.0 on macOS 12.7
Firefox 124.0.1 on macOS 12.7
Safari 17.4.1 on macOS 12.7
Opera 109 on Windows 10
Edge 123 on Windows 10
Edge 109 on Windows 8 (latest supported on that OS)

# Mobile
Firefox on Pixel 4a (vert & horiz)
Chrome on Pixel 4a (vert & horiz)

**Review by srjfoo — APPROVED** [review]

**Comment by srjfoo** [comment]
### iPad running 17.3.1
All three of Safari, Chrome and Firefox look good.

**Comment by cpeel** [comment]
Last push merely squashed the PR commits into each other. Will merge after CI passes.
